### DISCUSSION TITLE
QC for Consent Assessment

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#203] [General]
# QC for Consent Assessment

- [ ] Documentation
    - [x] Function name captured in [Roxygen title](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block)(e.g. "Adverse Event Assessment")
    - [ ] Assessment Purpose captured in [Roxygen description](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."
    - [ ] Input Data Requirements are described in the the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Data section (`#' @details # Data Requirements`).
    - [ ] Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Statistics section (`#' @details # Statistical Assumptions`).
    - [ ] Parameters have detailed documentation using `@param`. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)
    - [ ] All external dependencies are captured. Use `@importFrom` when importing 3 or fewer functions, and use `@import` otherwise. 
    - [x] Output Data Standards are described under `@returns`
    - [x] At least 1 example is provided under `@examples`
- [ ] Error Checking
    - [ ]  Basic checks for all all parameters should be included using `stopifnot()` or similar logic. (e.g. `stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput))`)
    - [ ] Tests confirm that `stopifnot()` parameter checks are working as expected. 
    - [ ] Tests confirm that the input data has required columns (if any)
    - [ ] Tests confirm that the output data has the expected columns
    - [ ] Tests confirm intended functionality for each parameter
    - [ ] Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed. 
- [ ] Basic QC
    - [ ] Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing. 
    - [x] Code is well commented and easy to read
    - [x] No file paths or other company-specific data is present
    - [x] Function calls from non-tidyverse dependencies are called via `::`
    - [x] devtools::check() passes with no errors/warnings/notes
    - [x] package docs up to date. devtools::document() doesn't change any files
    - [x] Uses tidyverse best practices for standard data manipulation (start discussion thread if unclear)
    - [x] All new dependencies add significant value (start discussion thread)
    - [x] All GitHub Actions run with no errors

**Comment by mattroumaya** [comment]
Grouping this one by function instead of by line item in the QC checklist. 

cc: @jwildfire - after your review I'll file separate issues to address in `v0.2.0`

## Consent_Assess()

**Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
- Contains [TODO on line 5](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L5) 

**Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
- There is no statistics section - unsure if needed though, because only doing flagging with thresholds

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- [`nThreshold` description](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L22) should probably be updated - is worded somewhat awkwardly (_"nThreshold integer threshold values Flagging, integer values greater than will be flagged."_)
- [`nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L22) - need to declare default value of 0.5
- Minor - add or omit periods to be consistent

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- [`nThreshold` is missing check](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L45-L50) for `as.numeric()`/`as.integer()` and not `NA`. 

**Tests confirm that stopifnot() parameter checks are working as expected.**
- [Need tests for `nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L45-L50)

**Tests confirm intended functionality for each parameter**
- [Need tests for `nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Assess.R)

**Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
- Need [tests](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Assess.R) for `NA` values in columns: `SubjectID, SiteID, Count`

## Consent_Map_Raw()

**Input Data Requirements are described in the the Roxygen details under a Data section (#' @details # Data Requirements).**
- [`RandData` - Randomization Date](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R#L21) is not spelled correctly. Should be updated to `RandDate`.

**All external dependencies are captured. Use @importFrom when importing 3 or fewer functions, and use @import otherwise.**
- Is [`@import lubridate`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R) needed? I don't see where it's used. 

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- "SubjectID, SiteID and RandData column not found in dfIxrsrand" is [incorrect](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R#L47)
  - change `RandData` to `RandDate`
  - change `dfIxrsrand` to `dfRDSL`
- Need to check that `strConsentReason` is length 1
- Need to check that `strConsentReason` is character
- Do we need to check `NA` values in `SubjectID` like we do in [`AE_Map_*`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Map_Raw.R#L43)?

**Tests confirm that stopifnot() parameter checks are working as expected.**
- Need clarification for `NA` handling as noted above

**Tests confirm that the input data has required columns (if any)**
- Need tests for `dfRDSL` columns: `SubjectID`, `SiteID`, `RandDate`
  - To be consistent with `dfConsent` tests, suggest:

```r
  expect_error(
    Consent_Map_Raw(
      dfConsent_test,
      dfRDSL = dfRDSL_test %>% select(- SubjectID)
    )
  )

  expect_error(
    Consent_Map_Raw(
      dfConsent_test,
      dfRDSL = dfRDSL_test %>% select(- SiteID)
    )
  )

  expect_error(
    Consent_Map_Raw(
      dfConsent_test,
      dfRDSL = dfRDSL_test %>% select(- RandDate)
    )
  )
```

**Tests confirm intended functionality for each parameter**
- Need [test(s)](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Map_Raw.R) for `strConsentReason`

**Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
- Need [tests](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Map_Raw.R) for `NA` values in all required columns for both datasets (`dfConsent` & `dfRDSL`)

**Reply by jwildfire** [reply]
Looks good @mattroumaya. Let's go ahead and file one issue to clean up the items with 👍  below. 

> ## Consent_Assess()
> **Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
> 
> * Contains [TODO on line 5](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L5)

Added note in #83. No more action needed for now. 

> **Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
> 
> * There is no statistics section - unsure if needed though, because only doing flagging with thresholds

👍  Would be good to add a section just to mention that no formal modeling is done for this assessment. 

> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
> 
> * [`nThreshold` description](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L22) should probably be updated - is worded somewhat awkwardly (_"nThreshold integer threshold values Flagging, integer values greater than will be flagged."_)
> * [`nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L22) - need to declare default value of 0.5
> * Minor - add or omit periods to be consistent

👍 

> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * [`nThreshold` is missing check](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L45-L50) for `as.numeric()`/`as.integer()` and not `NA`.

👍 

 
> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * [Need tests for `nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Assess.R#L45-L50)

👍 
 
> **Tests confirm intended functionality for each parameter**
> 
> * [Need tests for `nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Assess.R)
>
> **Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
 
👍 

> * Need [tests](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Assess.R) for `NA` values in columns: `SubjectID, SiteID, Count`

👍

**Reply by jwildfire** [reply]
Good to create issue.

I also want to discuss the possibility of using a more generalized approach to checking input data. Feel like we're going to end up creating a lot of duplicated logic for checking input data in map_raw across assessments. Not sure if there is a good way to avoid that, but probably worth discussing.  

> ## Consent_Map_Raw()
> **Input Data Requirements are described in the the Roxygen details under a Data section (#' @details # Data Requirements).**
> 
> * [`RandData` - Randomization Date](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R#L21) is not spelled correctly. Should be updated to `RandDate`.

👍 

> **All external dependencies are captured. Use @importFrom when importing 3 or fewer functions, and use @import otherwise.**
> 
> * Is [`@import lubridate`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R) needed? I don't see where it's used.

👍  It's from a previous iteration. We can remove. 
 
> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * "SubjectID, SiteID and RandData column not found in dfIxrsrand" is [incorrect](https://github.com/Gilead-BioStats/gsm/blob/dev/R/Consent_Map_Raw.R#L47)
>   
>   * change `RandData` to `RandDate`
>   * change `dfIxrsrand` to `dfRDSL`
> * Need to check that `strConsentReason` is length 1
> * Need to check that `strConsentReason` is character
> * Do we need to check `NA` values in `SubjectID` like we do in [`AE_Map_*`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Map_Raw.R#L43)?

👍 Yes. Let's add the same logic for `SubjectID`

> 
> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * Need clarification for `NA` handling as noted above

👍 

> **Tests confirm that the input data has required columns (if any)**
> 
> * Need tests for `dfRDSL` columns: `SubjectID`, `SiteID`, `RandDate`
>   
>   * To be consistent with `dfConsent` tests, suggest:
> 
> ```r
>   expect_error(
>     Consent_Map_Raw(
>       dfConsent_test,
>       dfRDSL = dfRDSL_test %>% select(- SubjectID)
>     )
>   )
> 
>   expect_error(
>     Consent_Map_Raw(
>       dfConsent_test,
>       dfRDSL = dfRDSL_test %>% select(- SiteID)
>     )
>   )
> 
>   expect_error(
>     Consent_Map_Raw(
>       dfConsent_test,
>       dfRDSL = dfRDSL_test %>% select(- RandDate)
>     )
>   )
> ```

👍 

> **Tests confirm intended functionality for each parameter**
> 
> * Need [test(s)](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Map_Raw.R) for `strConsentReason`

👍 
 
> **Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
> 
> * Need [tests](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_Consent_Map_Raw.R) for `NA` values in all required columns for both datasets (`dfConsent` & `dfRDSL`)

👍

**Comment by jwildfire** [comment]
Looks good @mattroumaya. Go ahead and create issues for v0.2 for each function as mentioned in the replies above. We're still good to proceed with v0.1 release.
